src/lib/state.ts

```diff
@@ -0,0 +1,277 @@
+      let stateAsFields = this._contract.stateType.toFields(state);
+      let party = this._contract.instance.self;
+
+      stateAsFields.forEach((x, i) => {
+        party.body.preconditions.account.state[layout.offset + i].isSome =
+          Bool(true);
+        party.body.preconditions.account.state[layout.offset + i].value = x;
+      });
+    },
+
+    get() {
+      if (this._contract === undefined)
+        throw Error(
+          'get can only be called when the State is assigned to a SmartContract @state.'
+        );
+      let layout = getLayoutPosition(this._contract);
+      let address: PublicKey = this._contract.instance.address;
+      let stateAsFields: Field[];
+      let inProver = GlobalContext.inProver();
+      if (!GlobalContext.inCompile()) {
+        let account: Account;
+        try {
+          account = Mina.getAccount(address);
+        } catch (err) {
+          // TODO: there should also be a reasonable error here
+          if (inProver) {
+            throw err;
+          }
+          throw Error(
+            `${this._contract.key}.get() failed, because the zkapp account was not found in the cache. ` +
+              `Try calling \`await fetchAccount(zkappAddress)\` first.`
+          );
+        }
+        if (account.zkapp === undefined) {
+          // if the account is not a zkapp account, let the default state be all zeroes
+          stateAsFields = Array(layout.length).fill(Field.zero);
+        } else {
+          stateAsFields = [];
+          for (let i = 0; i < layout.length; ++i) {
+            stateAsFields.push(account.zkapp.appState[layout.offset + i]);
+          }
+        }
+        // in prover, create a new witness with the state values
+        // outside, just return the state values
+        stateAsFields = inProver
+          ? Circuit.witness(
+              Circuit.array(Field, layout.length),
+              () => stateAsFields
+            )
+          : stateAsFields;
+      } else {
+        // in compile, we don't need the witness values
+        stateAsFields = Circuit.witness(
+          Circuit.array(Field, layout.length),
+          (): Field[] => {
+            throw Error('this should never happen');
+          }
+        );
+      }
+      let state = this._contract.stateType.ofFields(stateAsFields);
+      this._contract.stateType.check?.(state);
+      return state;
+    },
+
+    async fetch() {
+      if (this._contract === undefined)
+        throw Error(
+          'fetch can only be called when the State is assigned to a SmartContract @state.'
+        );
+      if (Mina.currentTransaction !== undefined)
+        throw Error(
+          'fetch is not intended to be called inside a transaction block.'
+        );
+      let layout = getLayoutPosition(this._contract);
+      let address: PublicKey = this._contract.instance.address;
+      let { account } = await fetchAccount(address);
+      if (account === undefined) return undefined;
```

Comment:
Should we throw here? If you try to call set and it returned undefined here we’ll throw and it will just be more confusing

Reply:
👍